Feature/decouple okta

jppdpf opened the pull request.

jppdpf committed:
Update client_secrets.json.dist

Updated URIs so that the OIDC has a base uri and the other endpoints are referenced with relative URIs.

jppdpf committed:
Update helpers.py

Added Configuration class to perform validation on the configuration.

jppdpf committed:
Update helpers.py

Updated all variables to private.
Added property decorators.

jppdpf committed:
Decoupled Okta from project. Created services. Changed configuration file.

Updated client_secrets.json.dist.
Renamed configuration file.
Moved helpers and form to created utils folder.
Updated templates to remove the Okta static references.
Updated Page Title to remove Okta static reference.
Created services package.
Moved Provider, OktaProvider and OneLoginProvider class to services.providers package.
Created yoonik service class.
Moved all api requests in main to the associated service class.
Updated delete yoonik account endpoint, so that after its deleted the user is logged out.

jppdpf committed:
Improved code quality. Rename configuration file.

Renamed client_secrets configuration file to config.
Updated gitignore entry for the configuration file.
Pylint improvements.
Renamed user to user_service.
Renamed yoonik to yoonik_service.

jppdpf committed:
Updated project to use yk-utils.

Refactored code.
Updated web UI text.
Updated configuration json file.
Added OneLogin and Okta integration documentation.

jppdpf committed:
Merge branch 'main' into feature/decouple-okta

# Conflicts:
#	README.md
#	client_secrets.json.dist
#	docs/index.md
#	main.py
#	requirements.txt

jppdpf commented on services/providers/okta_provider.py:
```diff
@@ -0,0 +1,38 @@
+""" Okta OIDC Provider """
+import asyncio
+import requests
+from okta_jwt_verifier import IDTokenVerifier
```
This needs to be removed.

vitormcpedro commented on README.md:
```diff
@@ -1,24 +1,36 @@
+# OIDC Hosted Login + YooniK Face Authentication 
 
-This example shows you how to use Flask to log in to your application with an Okta Hosted Login page.
+## Supported OIDCs
```
Supported OIDC **providers**

vitormcpedro commented on README.md:
```diff
@@ -1,24 +1,36 @@
+3. User is redirected back to this application with an access code that is then exchanged for an access token. 
+4. Token is validated.
+5. User takes a selfie to perform a second-factor face authentication with YooniK.
+6. User is logged.
```
User is logged **in**.

vitormcpedro commented on README.md:
```diff
@@ -32,49 +44,34 @@ Then install dependencies:
 
-* **Client ID** and **Client Secret** - These can be found on the **Sign On** tab of the YooniK app integration that you installed earlier in the Okta Admin Console. 
-* **Open ID Connect URLs** - These are the **authorization_endpoint**, **token_endpoint** and **userinfo_endpoint** for your Okta domain that can be found by clicking on **OpenID Provider Metadata** link under the **Sign On** tab.
+[Okta](docs/okta.md) and [OneLogin](docs/onelogin.md) configuration are available in this repository in the `.\docs` folder.
```
...configuration **instructions** are available...

vitormcpedro commented on README.md:
```diff
@@ -32,49 +44,34 @@ Then install dependencies:
+### Known issues
 
-**Note:** If you are currently using your Developer Console, you already have a Single Sign-On (SSO) session for your Org. You will be automatically logged into your application as the same user that is using the Developer Console. You may want to use an incognito tab to test the flow from a blank slate.
+> If you already have a Single Sign-On session to you provider, you may want to use your browser in incognito mode to test the flow from a blank slate. 
```
...session **with your OIDC** provider...

vitormcpedro commented on config.json.dist:
```diff
@@ -0,0 +1,20 @@
+{
+  "oidc_base_url": "yourOidcBaseUrl",
+  "issuer_url": "yourIssuerUrl",
```
This is not used

vitormcpedro commented on docs/okta.md:
```diff
@@ -0,0 +1,115 @@
+
+## Overview
+
+This document is intended to be use with [YooniK Sample App](https://github.com/dev-yoonik/yoonik-okta-example-python), in which the app login is performed 
```
We can use the same overview text as before. All sections from this document should be kept as before because they follow Okta requirements.

vitormcpedro commented on services/providers/provider.py:
```diff
@@ -0,0 +1,82 @@
+        """
+        Requests an access token from the providers token endpoint.
+        :param code: the code received from the provider
+        :param redirect_url: the redirect url in which
```
incomplete description

vitormcpedro commented on services/providers/okta_provider.py:
```diff
@@ -0,0 +1,38 @@
+
+class OktaProvider(Provider):
+    """ Okta OIDC Provider Class """
+    def __init__(self, config, app_state='ApplicationState', nonce='SampleNonce'):
```
add typing

vitormcpedro commented on services/providers/okta_provider.py:
```diff
@@ -0,0 +1,38 @@
+        self.__app_state = app_state
+        self.__nonce = nonce
+
+    def get_login_url(self, oidc_scopes=None) -> str:
```
add typing

vitormcpedro commented on services/providers/okta_provider.py:
```diff
@@ -0,0 +1,38 @@
+
+        return request_url
+
+    def is_token_valid(self, token):
```
add typing

vitormcpedro commented on services/providers/onelogin_provider.py:
```diff
@@ -0,0 +1,46 @@
+class OneLoginProvider(Provider):
+    """ OneLogin OIDC Provider Class """
+
+    def __init__(self, config):
```
add typing

vitormcpedro commented on services/providers/onelogin_provider.py:
```diff
@@ -0,0 +1,46 @@
+        self.name = "OneLogin"
+        self._token_validation_url = config.token_validation_url
+
+    def get_login_url(self, oidc_scopes=None):
```
add typing

vitormcpedro commented on services/providers/onelogin_provider.py:
```diff
@@ -0,0 +1,46 @@
+        request_url = f"{self._auth_url}?{requests.compat.urlencode(query_params)}"
+        return request_url
+
+    def is_token_valid(self, token):
```
add typing

vitormcpedro commented on services/providers/provider.py:
```diff
@@ -0,0 +1,82 @@
+class Provider(ABC):
+    """ Provider base class """
+
+    def __init__(self, config, default_scopes=None):
```
add typing

vitormcpedro commented on services/providers/provider.py:
```diff
@@ -0,0 +1,82 @@
+        self._default_scopes = default_scopes
+
+    @abstractmethod
+    def is_token_valid(self, token) -> bool:
```
add typing

vitormcpedro commented on services/providers/provider.py:
```diff
@@ -0,0 +1,82 @@
+        ...
+
+    @abstractmethod
+    def get_login_url(self, oidc_scopes) -> str:
```
add typing

vitormcpedro commented on services/providers/provider.py:
```diff
@@ -0,0 +1,82 @@
+        """
+        ...
+
+    def get_token(self, code, redirect_url):
```
Add typing. Maybe a better naming is "request_access_token" to not confuse with id token.

vitormcpedro commented on services/providers/provider.py:
```diff
@@ -0,0 +1,82 @@
+
+        return response["access_token"]
+
+    def get_user_info(self, token):
```
Add typing

vitormcpedro commented on services/user_service.py:
```diff
@@ -1,11 +1,12 @@
 
 
-class User(UserMixin):
+class UserService(UserMixin):
```
This is not a Service. It is data model for the User. We can keep the name as before.

vitormcpedro commented on services/user_service.py:
```diff
@@ -21,8 +22,20 @@ def claims(self):
     def create(user_id, name, email):
-        USERS_DB[user_id] = User(user_id, name, email)
+        """
+        Create a new user with the specified
```
incomplete description

vitormcpedro commented on templates/home.html:
```diff
@@ -1,21 +1,26 @@
+        <li>Authenticate with YooniK.</li>
+    </ol>
+    <p>That's it! After this steps you will be logged in this app.</p>
+    <p><b>Note: an YooniK account will be created with your email. Once logged in if you choose, you may delete the account.</b></p>
```
The YooniK account does not use the user e-mail, it uses the "sub" (an id)

vitormcpedro commented on utils/config.py:
```diff
@@ -0,0 +1,56 @@
+        :param config: configuration data
+        :return:
+        """
+        self.oidc_base_url = get_url(config["oidc_base_url"])
```
This line does not perform any action. Maybe we just want to call **is_valid_url**?

vitormcpedro commented on utils/config.py:
```diff
@@ -0,0 +1,56 @@
+        """
+        self.oidc_base_url = get_url(config["oidc_base_url"])
+
+        domain = get_url_domain(self.oidc_base_url).lower()
```
As referenced in yk-utils PR, i believe it is cleaner to use the hostname here.

vitormcpedro commented on utils/config.py:
```diff
@@ -0,0 +1,56 @@
+class Configuration:
+    """ Configuration Class """
+
+    def __init__(self, config_file="./config.json"):
```
Add typing

vitormcpedro commented on utils/config.py:
```diff
@@ -0,0 +1,56 @@
+
+        domain = get_url_domain(self.oidc_base_url).lower()
+        if domain in ('okta', 'onelogin'):
+            self.provider_name = domain
```
You can remove **self.name** from Provider classes and use **self.provider_name**, right? To avoid redundant data.

vitormcpedro commented on utils/config.py:
```diff
@@ -0,0 +1,56 @@
+        self.token_url = get_url(self.oidc_base_url, config["token_url"])
+        self.user_info_url = get_url(self.oidc_base_url, config["userinfo_url"])
+
+        self.redirect_url = get_url(config["redirect_url"])
```
This line does not perform any action. Maybe we just want to call is_valid_url?

vitormcpedro commented on utils/config.py:
```diff
@@ -0,0 +1,56 @@
+        self.client_id = config["client_id"]
+        self.client_secret = config["client_secret"]
+
+        self.yk_authentication_url = get_url(config["yoonik_authentication_api_url"])
```
This line does not perform any action. Maybe we just want to call is_valid_url?

jppdpf replied:
I called the get_url because it already validates the url and raise the exception if it's in a bad format.

jppdpf committed:
Updated with review critics.

Removed issuer url from documentation, configuration file and configuration class.
Updated okta documentation to contain the maximum information possible of the previous doc.
Updated readme.
Reverted User->UserService change.
Removed unnecessary imports from packages that are not required anymore.
Added docstrings and typings.
Updated with yk-utils reviews.

jppdpf committed:
Updated User class.

Renamed user_service to user.
Created data_model and moved the user file.
